wordpress-mobile/WordPress-Android
# Build gutenberg-mobile locally without JitPack

We currently experience high failure rate on JitPack building the gutenberg-mobile artifact for WPAndroid, so this PR uses the "JITPACK" mode in gutenberg-mobile to launch Node/Yarn to build the block editor's JS bundle and embed it in the gutenberg-mobile bridge artifact.

No need to install Node or Yarn, that's done automatically but the [android-maven-gradle-plugin](https://github.com/dcendents/android-maven-gradle-plugin) Gradle plugin in a temporary folder.

Related gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2231

## Changes in this PR

1. Remove dependency to the JitPack-built gutenberg-mobile library artifact

2. Check out the full (albeit shallow) gutenberg-mobile git submodule, including its sub-submodules. Needed for building the gutenberg-mobile native and JS code locally.

3. Have a dedicated CircleCI job (`gutenberg-bundle-build`) to build the JS bundle of gutenberg-mobile. Other build jobs depend on this job to finish first. The JS bundle is stored in a CircleCI workspace to be picked up by the other jobs.

4. Locally, the JS bundle task is called always. It's suppressed on CI by using an env flag.

5. The JS bundle job is using a nested Node+Yarn toolchain, managed via gradle. No need to manually install the tools, and they are not installed globally either.

6. Added quite a few of lint checks to the "ignore list" (lint-baseline.xml). Those violations should be wrangled separately, inside the gutenberg-mobile repo.

7. Added the `gutenberg-mobile` regex to the checkstyle ignore list for the same reason.

~8. Bumped the gradle daemon JVM mem to 3g and disabled the `parallel` option too, to help the jobs succeed on CI, although that was an experiment for when the JS bundling was performed as part of each gradle build. Probably not needed now that the JS bundle has its own job, and will have to revert the change.~

## Known issues

1. The `yarn_install` gradle task is running on every build, even though the gutenberg-mobile artifact doesn't need to be built again. This should be fixed with a better tasks dependency configuration.
2. There are build artifacts left behind in the libs/gutenberg-mobile/bundle/android/raw/ folder, that are not clear yet what to do with (git ignore them or something else). Also, the react-native-gutenberg-bridge/android/src/main/assets/ folder is also not handled in VSC yet.

### To test A:

Run the app and launch the block editor, it should work normally.

### To test B:

Check out the "Builds" tab in https://jitpack.io/#wordpress-mobile/gutenberg-mobile and ensure that there's no build for `d74a56` listed, the gutenberg-mobile hash currently pointed to. 

PR submission checklist:
